File: moq-web/src/internal/message/track.ts
import type { Reader, Writer } from "@okdaichi/golikejs/io";
import { MessageDecoder, MessageEncoder, readFull, readVarint } from "./message.ts";

export interface TrackMessageInit {
	broadcastPath?: string;
	trackName?: string;
}

/**
 * TRACK message, sent by a subscriber as the first message on a Track
 * Stream to request a track's immutable publisher properties.
 *
 * ```text
 * TRACK Message {
 *   Message Length (i)
 *   Broadcast Path (s)
 *   Track Name (s)
 * }
 * ```
 */
export class TrackMessage {
	broadcastPath: string;
	trackName: string;

	constructor(init: TrackMessageInit = {}) {
		this.broadcastPath = init.broadcastPath ?? "";
		this.trackName = init.trackName ?? "";
	}

	/**
	 * Encodes the message to the writer.
	 */
	async encode(w: Writer): Promise<Error | undefined> {
		const e = new MessageEncoder();
		e.string(this.broadcastPath);
		e.string(this.trackName);
		const [, err] = await w.write(e.frame());
		return err;
	}

	/**
	 * Decodes the message from the reader.
	 */
	async decode(r: Reader): Promise<Error | undefined> {
		const [msgLen, , err1] = await readVarint(r);
		if (err1) return err1;

		const buf = new Uint8Array(msgLen);
		const [, err2] = await readFull(r, buf);
		if (err2) return err2;

		const d = new MessageDecoder(buf);
		this.broadcastPath = d.string();
		this.trackName = d.string();

		return undefined;
	}
}

File: moqt/internal/message/track.go
package message

import (
	"io"
)

/*
 *	TRACK Message {
 *	  Message Length (i)
 *	  Broadcast Path (s)
 *	  Track Name (s)
 *	}
 */
type TrackMessage struct {
	BroadcastPath string
	TrackName     string
}

func (tm TrackMessage) Len() int {
	return StringLen(tm.BroadcastPath) + StringLen(tm.TrackName)
}

func (tm TrackMessage) Encode(w io.Writer) error {
	msgLen := tm.Len()
	b := make([]byte, 0, msgLen+VarintLen(uint64(msgLen)))

	b, _ = WriteMessageLength(b, uint64(msgLen))
	b, _ = WriteString(b, tm.BroadcastPath)
	b, _ = WriteString(b, tm.TrackName)

	_, err := w.Write(b)

	return err
}

func (tm *TrackMessage) Decode(src io.Reader) error {
	size, err := ReadMessageLength(src)
	if err != nil {
		return err
	}

	if size > MaxMessageSize {
		return ErrMessageTooLarge
	}

	b := make([]byte, size)

	_, err = io.ReadFull(src, b)
	if err != nil {
		return err
	}

	str, n, err := ReadString(b)
	if err != nil {
		return err
	}
	tm.BroadcastPath = str
	b = b[n:]

	str, n, err = ReadString(b)
	if err != nil {
		return err
	}
	tm.TrackName = str
	b = b[n:]

	if len(b) != 0 {
		return ErrMessageTooShort
	}

	return nil
}
